src/System.Net.Security/src/System/Net/Security/SslState.cs

```diff
@@ -1346,25 +1358,34 @@ internal void FinishWrite()
 
             lock (this)
             {
-                object obj = _queuedWriteStateRequest;
-                if (obj == null)
-                {
-                    // A repeated call.
-                    return;
-                }
+                HandleQueuedCallback(ref _queuedWriteStateRequest);
+            }
+        }
 
-                _queuedWriteStateRequest = null;
-                if (obj is LazyAsyncResult)
-                {
-                    // Sync handshake is waiting on other thread.
-                    ((LazyAsyncResult)obj).InvokeCallback();
-                }
-                else
-                {
-                    // Async handshake is pending, start it on other thread.
-                    // Consider: we could start it in on this thread but that will delay THIS write completion
+        private void HandleQueuedCallback(ref object queuedStateRequest)
+        {
+            if(queuedStateRequest == null)
+            {
+                return;
+            }
+                   
+            object obj = queuedStateRequest;
+            queuedStateRequest = null;
+            switch (obj)
+            {
+                case LazyAsyncResult lazy:
+                    lazy.InvokeCallback();
+                    break;
+                case TaskCompletionSource<int> tsc when tsc.Task.AsyncState != null:
+                    Memory<byte> array = (Memory<byte>)tsc.Task.AsyncState;
+                    tsc.SetResult(CheckOldKeyDecryptedData(array));
+                    break;
+                case TaskCompletionSource<int> tsc:
+                    tsc.SetResult(0);
+                    break;
+                default:
                     ThreadPool.QueueUserWorkItem(new WaitCallback(AsyncResumeHandshake), obj);
```

Comment:
I think I already asked this, but... this is rare, right?

Reply:
Super, if at all it happens 1x per connection and that is only if you are using client certs and only if the server decides it needs a different client cert to the one you negotiated in the handshake. Then you have to attempt to read or write while you negotiate that second cert.

Also this functionality is not even allowed to happen during an HTTP/2 connection so it will become even rarer because if your site will ever use HTTP/2 you can't rely on it for HTTP.

This whole re-handshake is ugly.. I would love to change it, in fact I have prototyped sucessfully switching to a semaphore slim and making the handshake async as well. But I am concerned about the lack of tests here.....

TLDR : it basically never happens, when it does you have to wait for a full handshake, so perf is killed by this point anyway.